tensorflow/core/grappler/optimizers/auto_mixed_precision_lists.h

```diff
@@ -0,0 +1,297 @@
+    };
+    UpdateList(&list, to_add, to_remove);
+    return list;
+  }
+
+  // Returns the set of ops that are considered numerically-safe (for execution
+  // in fp16), but which may be made unsafe by an upstream blacklist op.
+  static std::set<string> GrayList() {
+    if (IsPseudoFastMath()) {
+      return std::set<string>{};
+    }
+    string to_add, to_remove;
+    ReadStringFromEnvVar("TF_AUTO_MIXED_PRECISION_GRAPH_REWRITE_GRAYLIST_ADD",
+                         "", &to_add);
+    ReadStringFromEnvVar(
+        "TF_AUTO_MIXED_PRECISION_GRAPH_REWRITE_GRAYLIST_REMOVE", "",
+        &to_remove);
+
+    auto list = std::set<string>{
+        "Add",
+        "AddN",
+        "AddV2",
+        "AvgPool",
+        "AvgPool3D",
+        "AvgPool3DGrad",
+        "AvgPoolGrad",
+        "BiasAdd",
+        "BiasAddGrad",
+        "BiasAddV1",
+        "Elu",
+        "EluGrad",
+        "Erf",
+        "Erfc",
+        "FloorDiv",
+        "FusedBatchNormV2",
+        "FusedBatchNormGradV2",
+        "Inv",
+        "LeakyRelu",
+        "LeakyReluGrad",
+        "Mul",
+        "Prod",
+        "RealDiv",
+        "Reciprocal",
+        "Sigmoid",
+        "SigmoidGrad",
+        "Softplus",
+        "SoftplusGrad",
+        "Sqrt",
+        "Sub",
+        "Tanh",
+        "TanhGrad",
+    };
+    UpdateList(&list, to_add, to_remove);
+    return list;
+  }
+
+  // Returns the set of ops that are considered numerically-dangerous (i.e.,
+  // unsafe for execution in fp16) and whose effects may also be observed in
+  // downstream nodes (e.g., in Exp -> Add, the Add is unsafe due to the Exp).
+  static std::set<string> BlackList() {
+    if (IsPseudoFastMath()) {
+      return std::set<string>{};
+    }
+    string to_add, to_remove;
+    ReadStringFromEnvVar("TF_AUTO_MIXED_PRECISION_GRAPH_REWRITE_BLACKLIST_ADD",
+                         "", &to_add);
+    ReadStringFromEnvVar(
+        "TF_AUTO_MIXED_PRECISION_GRAPH_REWRITE_BLACKLIST_REMOVE", "",
+        &to_remove);
+
+    auto list = std::set<string>{
+        "Exp",
+        "Expm1",
+        "L2Loss",
+        "Log",
+        "Log1p",
+        "LogSoftmax",
+        "Mean",
+        "Pow",
+        "SaveV2",
+        "Softmax",
+        "SoftmaxCrossEntropyWithLogits",
+        "SparseSoftmaxCrossEntropyWithLogits",
+        "Sum",
+    };
+    UpdateList(&list, to_add, to_remove);
+    return list;
+  }
+
+  // Returns the set of ops that do not have numerically-significant effects
+  // (i.e., they are always considered safe for execution in fp16 precision).
+  static std::set<string> ClearList() {
+    if (IsPseudoFastMath()) {
+      return std::set<string>{};
+    }
+    string to_add, to_remove;
+    ReadStringFromEnvVar("TF_AUTO_MIXED_PRECISION_GRAPH_REWRITE_CLEARLIST_ADD",
+                         "", &to_add);
+    ReadStringFromEnvVar(
+        "TF_AUTO_MIXED_PRECISION_GRAPH_REWRITE_CLEARLIST_REMOVE", "",
+        &to_remove);
+
+    auto list = std::set<string>{
```

Comment:
I'm slightly worried that one may accidentally add a stateful op without realizing they also need to call AddDataStructureOpsToMap. Can you add a comment somewhere in this function, something like:

Note: if a stateful op is added, such as StackPopV2, is added to the clearlist, you must also modify AutoMixedPrecisionImpl to call AddDataStructureOpsToMap() with that op.

Reply:
Added a comment to that effect